src/nvim/api/extmark.c

```diff
@@ -184,10 +190,45 @@ static Array extmark_to_array(const ExtmarkInfo *extmark, bool id, bool add_dict
       PUT(dict, "virt_lines_leftcol", BOOLEAN_OBJ(virt_lines_leftcol));
     }
 
-    if (decor->hl_id || kv_size(decor->virt_text) || decor->ui_watched) {
+    if (decor->sign_text) {
+      PUT(dict, "sign_text", STRING_OBJ(cstr_to_string(decor->sign_text)));
+    }
+
+    // uncrustify:off
+
+    struct { const char *name; const int val; } hls[] = {
+      { "sign_hl_group"      , decor->sign_hl_id       },
+      { "number_hl_group"    , decor->number_hl_id     },
+      { "line_hl_group"      , decor->line_hl_id       },
+      { "cursorline_hl_group", decor->cursorline_hl_id },
+      { NULL, 0 },
+    };
+
+    // uncrustify:on
+
+    for (int j = 0; hls[j].name && hls[j].val; j++) {
+      if (hls[j].val) {
+        String name = cstr_to_string((const char *)syn_id2name(hls[j].val));
```

Comment:
same here. hl ids are already an exposed concept, so just returning the `hl_id` without converting to string is fine.

Reply:
or an opts flag. 

Regardless, new fields for highlights should use hl_id only. the precedence of some unnecessary conversions is not a reason to introduce _yet more_ costly conversions all over the API. ("consistency" should not mean cargo-cult copying past mistakes)

Reply:
Opts flag is a good idea. Perhaps `hl_names = true`, and defaults to `false`.

Reply:
Defaulting to false would break backwards compatibility. The opts flag would work but if we want to promote the `false` default we would still need a `get_extmarks2`.

Reply:
@bfredl what do you think.

Reply:
> fewer string allocations

I.e. if we value this we could also opt to set a boolean `virt_text_pos_eol/overlay/win_col/right_align"` to true rather than `virt_text_pos = "string"`. Same for `hl_mode`. That would leave only `text` elements to be actual strings.

Reply:
No keep enums as strings. Lua will intern them, so we just need to maintain them as constants in the backend.

Reply:
We are copying the string in the backend though (because script values are free'd?).

Reply:
OK, just need to know which route to take this PR in. `opts->hl_names` vs `opts->details2` vs `get_extmarks2` @bfredl.

Reply:
For now let's do neither and add `nvim_buf_get_extmarks2` in the future. For this PR just add all the hl id's and leave existing fields as they are.

Reply:
Do we really want to mix hl ids and group names in the details array then? Seems better to just decide what to do before acting.

Reply:
Ok, default to names and add an option for id's

Reply:
OK, added an `opts->verbose` flag to move forward. Although I still feel like if the intention is to add a `get_extmarks2` in the future because of this addition, we might as well do that now.

Reply:
Not necessarily, as we're entering the feature freeze for 0.9, and reworking this PR would mean we'd have to delay it until after the release.

Reply:
(Also, I think we've already added enough `nvim_foo2` APIs for one release ;))

Reply:
Yes that's what I did, but I thought we identified that we rather have the inverse but can't because of backwards compatibility.

Reply:
I don't mind too much at this point. I'd rather get something we can use for 0.9. We always have the possibility of introducing a new function now or in the future.

Do what you think is best.

Reply:
I don't mind either, we're discussing this because of @bfredl's original concern:

> hl ids are already an exposed concept (and accepted when setting extmarks) , so just returning the hl_id without converting to string is fine.

If @bfredl thinks this flag is useful I'm fine with adding it but if we rather just want to change the output of the details array unconditionally(and thus deprecate) I think we should postpone that and just go with the hl group names for now.